src/dinit-iostream.cc

```diff
@@ -0,0 +1,593 @@
+    // to this variable.
+    size_t pre_buffered_chars = 0;
+
+    if (count == 0) {
+        return 0; // Null/Empty message
+    }
+    else if (count > buf->get_size()) {
+        ssize_t r = flush_nx();
+        if (r < 0) {
+            // io_error was set by flush_nx() call
+            return -1;
+        }
+        ssize_t res = bp_sys::write(fd, msg, count);
+        if (res < 0) {
+            io_error = errno;
+            return -1;
+        }
+        return res;
+    }
+    else while (count > buf->get_free()) {
+        // If we haven't enough storage for caputring the message
+        // Firstly we try to fill buffer as much as possible and then
+        // write the buffer.
+        int prev_freespace = buf->get_free();
+        buf->append(msg, buf->get_free());
+        msg += prev_freespace;
+        pre_buffered_chars += prev_freespace;
+        count -= prev_freespace;
+        ssize_t r = flush_nx();
+        if (r < 0) {
+            // io_error was set by flush_nx()
+            return (pre_buffered_chars > 0) ? pre_buffered_chars : -1;
+        }
+    }
+
+    buf->append(msg, count);
+
+    return pre_buffered_chars + count;
+}
+
+bool ostream::open(const char *path) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY);
+    return (fd >= 0);
+}
+
+bool ostream::open(const char *path, const int flags) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY | flags);
+    return (fd >= 0);
+}
+
+bool ostream::open(const char *path, const int flags, const mode_t mode) noexcept
+{
+    fd = bp_sys::open(path, O_WRONLY | flags, mode);
+    return (fd >= 0);
+}
+
+bool ostream::close() noexcept
+{
+    if (flush_nx() < 0) return false; // Flushing failed
+    if (!is_open()) return false; // Closed already?
+    if (bp_sys::close(fd) < 0) return false; // Failed
+    fd = -1;
+    return true;
+}
+
+int ostream::current_state() noexcept
+{
+    int bits = 0;
+    if (buffer_failbit) bits |= io_states::buffer_failbit;
+    if (io_error) bits |= io_states::io_error;
+    return bits;
+}
+
+bool ostream::good() noexcept
+{
+    return (current_state() == 0);
+}
+
+bool ostream::buffer_failure() noexcept
+{
+    return (current_state() & buffer_failbit);
+}
+
+int ostream::io_failure() noexcept
+{
+    return io_error;
+}
+
+void ostream::clear() noexcept
+{
+    // Reset all failures
+    buffer_failbit = false;
+    io_error = 0;
+}
+
+ssize_t ostream::flush_nx() noexcept
+{
+    if (!buf) {
+        buffer_failbit = true;
+        return -1; // Tried to use output on null pointer buffer
+    }
+    ssize_t r = 0;
+    if (buf->get_length() > 0) {
+        char *ptr = buf->get_ptr(0);
+        unsigned len = buf->get_contiguous_length(ptr);
+        r = bp_sys::write(fd, ptr, len);
+
+        if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+        buf->consume(r);
+    }
+
+    if (buf->get_length() != 0) {
+        io_error = EAGAIN;
+        return -1;
+    }
```

Comment:
Looks like `write()` doesn't set errno on partial writing, I think setting `EAGAIN` is reasonable. @davmac314 WDYT?

Reply:
partial write isn't an error, that's why it doesn't set `errno`. If this function is intended to fully flush the buffer, it needs to try `write` again if there was only a partial write. (If the first write was partial due to a non-transient error, the second write will return the error anyway).

But also, the buffer is circular and as written this only flushes the first contiguous section.